test/testcommon/src/main/java/io/pravega/test/common/AssertExtensions.java

```diff
@@ -45,7 +45,11 @@
     public static <T> void assertEventuallyEquals(T expected, Callable<T> eval, long timeout) throws Exception {
         long endTime = System.currentTimeMillis() + timeout;
         while (endTime > System.currentTimeMillis()) {
-            if (expected == eval.call()) {
+            if (expected == null) {
```

Comment:
I think the logic you are trying to express here is more clearly stated as:

```
if ((expected == null && eval.call() == null) || (expected != null && expected.equals(eval.call())) {
    return;
}
```

Reply:
After looking at this method a few more times, I feel like we could move `testcommon.TestUtils.await(...)` in here and replace this method altogether. `await` is much more flexible and gives the caller the ability to implement their own condition check and fine-tune the check frequency and delay.

https://github.com/pravega/pravega/blob/master/test/testcommon/src/main/java/io/pravega/test/common/TestUtils.java#L68

Reply:
Yes, fixed with the compact condition in the AssertExtensions.java.

Reply:
I added awaitFirst method into TestUtils.java and started using it inside MetricsTest.
The reason is that await() test the condition and then await, however this is cache eviction test, if the object is accessed, it changes the expiration time.
The new awaitFirst turns the things around by await first and then check the condition.

Reply:
That's a good observation, @andreipaduroiu, but let me clarify what the proposal would be as I see a few options:

Option 1: we get rid of `assertEventuallyEquals` and only use `await`. That would require changing all occurrences of `assertEventuallyEquals`, which would be out of scope of this PR.
Option 2: we keep both `assertEventuallyEquals` and `await`.
   - Option 2a: we only replace `assertEventuallyEquals` with `await` in the test case we are fixing in this PR.
   - Option 2b: we replace more broadly the relevant occurrences of `assertEventuallyEquals`, but that again feels like out of scope of this PR.

Option 2a is acceptable, but then my next question is what happens to `assertEventuallyEquals`, do we fix in the way that is being proposed or leave it untouched? If the point that @kevinhan88 is raising is a concern, then we should be fixing it here in the case we keep `assertEventuallyEquals`.

What do you think?

Reply:
The scope of this PR is "Fix bug inside AssertExtensions". That is a bit broad, so I think option 1 can fall into scope. I don't want to have two different methods that are 90% similar and then we'll end up with people using both in the future. This has happened before and it was hard to fix afterwards.

There are only 17 callsites of this method. I think they can be resolved.

@kevinhan88 I don't see how using `await` will touch the cache in any other way compared to what you are doing now. You will simply pass your condition check as a `Supplier` to `await` which will have the same effect.

Reply:
@andreipaduroiu are these 2 methods for different purpose and usages? await() may return or throw timeout exception, while assertEventuallyEquals may return or throw AssertException. I don't see how one can replace another.

Reply:
- Both will fail the test.
- Both are designed to check a condition that is not necessarily true at the moment but may become true at one point in the future.
- `assertEventuallyEquals` can only check for equality. `await` is more flexible and allows you to verify other conditions (ex: not equals, greater, smaller, null, not null, etc.) 

- `await` is more efficient in that it doesn't call `System.currentTimeMillis` every time. 
- You can build `assertEventuallyEquals` using `await`, which will pass in the appropriate callback.

Reply:
I have to take responsibility for changing the PR title, I felt that the important part was fixing the bug in the method. As is, the title of the PR is broad so even option 1 falls into the scope of the PR title. The fix of the test case is a side-effect of the change. I'm ok with going with option 1 as long as @kevinhan88 is comfortable doing it in this PR.